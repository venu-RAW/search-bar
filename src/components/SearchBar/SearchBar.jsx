import React, { Component } from "react";
import styles from "./SearchBar.module.scss";
import PropTypes from "prop-types";

/**
 * Renders a <Searchbar /> component
 * @component
 *	<Searchbar
 *		searchData={this.state.data}
 *		result={this.showResult}
 *		searchKeys={["firstName", "lastName", "gender", "university"]}
 *		query={this.showQuery}
 *		placeholder="Search..."
 *		alignIcon="left"
 *	/>
 */

class SearchBar extends Component {
	state = {
		searchInput: "",
		resultArray: [],
	};

	/**
	 * @function handleChange
	 * @param {Event} event
	 ** Sets searchInput inside state and the searchInput value is passed to searchResult().
	 */
	handleChange = (event) => {
		const searchInput = event.target.value;
		const { query } = this.props;

		this.setState({
			searchInput,
		});

		query(searchInput);
		this.searchResult(searchInput);
	};

	/**
	 * @function searchResult
	 * @fires props.result
	 */
	searchResult = (searchInput) => {
		let resultArray = [];
		const { searchData, searchKeys } = this.props;

		if (searchInput.length)
			searchKeys.forEach((key) => {
				resultArray = searchData.filter((data) =>
					data[key].toLowerCase().includes(searchInput.toLowerCase())
				);
			});

		this.setState({
			resultArray,
		});

		/**
		 ** Execute callback named result and return it.
		 */
		return this.props.result(resultArray);
	};

	render() {
		const { placeholder, alignIcon } = this.props;

		return (
			<div
				data-test="searchBar"
				data-testid="searchBar"
				className={`${styles.searchBar} ${styles[alignIcon]}`}
			>
				<input
					data-testid="searchBarInput"
					type="text"
					className={styles.searchInput}
					value={this.state.searchInput}
					placeholder={placeholder}
					onChange={this.handleChange}
				/>
				<div className={styles.searchBtn}>
					<i className="fas fa-search"></i>
				</div>
			</div>
		);
	}
}

SearchBar.propTypes = {
	/**
	 ** The searchData type must be an array of objects ( Required ).
	 */
	searchData: PropTypes.array.isRequired,

	/**
	 ** The result type must funciton. It shows the result for the query.
	 */
	result: PropTypes.func.isRequired,

	/**
	 ** The query type must be a function. It shows the user query.
	 */
	query: PropTypes.func,

	/**
	 ** The keys on which you want to perform your search.
	 */
	searchKeys: PropTypes.arrayOf(PropTypes.string),

	/**
	 ** The placeholder type must be a string.
	 */
	placeholder: PropTypes.string,

	/**
	 ** The alignIcon type must be a string ( left or right )
	 ** By default is "right"
	 */
	alignIcon: PropTypes.oneOf(["left", "right"]),
};

SearchBar.defaultProps = {
	placeholder: "Search",
	alignIcon: "right",
};

export default SearchBar;
