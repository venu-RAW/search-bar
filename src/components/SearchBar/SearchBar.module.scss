$primary-color: #2c4466;
$secondary-color: #fff;

::placeholder {
	color: $secondary-color;
	opacity: 0.5;
}

.left {
	flex-direction: row-reverse;
}

.searchBar {
	position: absolute;
	top: 50%;
	left: 50%;
	transform: translate(-50%, -50%);
	height: 5rem;
	border-radius: 5rem;
	padding: 1rem 0.6rem;
	box-shadow: 0px 1px 5px #aaaaaa;
	background-color: $primary-color;
	display: flex;
	justify-content: center;
	align-items: center;

	&:hover > .searchInput {
		width: 30rem;
		padding: 0 1rem;
	}

	.searchInput {
		width: 0rem;
		height: 2.5rem;
		color: $secondary-color;
		font-weight: 600;
		font-size: inherit;
		letter-spacing: 1.5px;
		border: none;
		background: none;
		outline: none;
		transition: 0.5s;
		cursor: pointer;
	}

	.searchBtn {
		color: $secondary-color;
		width: 4rem;
		height: 4rem;
		border-radius: 100px;
		background-color: $primary-color;
		display: flex;
		justify-content: center;
		align-items: center;
		cursor: pointer;
	}
}
