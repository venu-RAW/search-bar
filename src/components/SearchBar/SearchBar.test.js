import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import SearchBar from "./SearchBar";

afterEach(cleanup);

describe("<SearchBar /> component", () => {
	let wrapper;

	let testProps = {
		searchData: [
			{
				id: 1,
				firstName: "Shannon",
			},
			{
				id: 2,
				firstName: "Worden",
			},
			{
				id: 3,
				firstName: "Idell",
			},
			{
				id: 4,
				firstName: "Shanti",
			},
		],
		result: jest.fn(),
		query: jest.fn(),
		searchKeys: ["firstName"],
		placeholder: "Search",
		alignIcon: "left",
	};

	beforeEach(() => {
		wrapper = render(<SearchBar {...testProps} />);
	});

	/*
	it("should render the SearchBar", () => {
		let { asFragment } = wrapper;
		expect(asFragment()).toMatchSnapshot();
	});
	*/

	it("should render the SearchBar Component", () => {
		let { getByTestId } = wrapper;

		expect(getByTestId("searchBar")).toBeTruthy();
	});

	it("should receive empty string", () => {
		let event = {
			target: {
				value: "",
			},
		};
		let { container, getByTestId } = wrapper;

		const searchBarInput = getByTestId("searchBarInput");

		const input = container.querySelector("input");
		fireEvent.change(input, event);

		expect(searchBarInput).not.toBeNull();
		expect(searchBarInput).toHaveValue("");
	});

	it("should receive an event, check length of resultArray", () => {
		let event = {
			target: {
				value: "S",
			},
		};
		let { container, getByTestId } = wrapper;

		const searchBarInput = getByTestId("searchBarInput");

		const input = container.querySelector("input");
		fireEvent.change(input, event);

		expect(searchBarInput).not.toBeNull();
		expect(searchBarInput).toHaveValue("S");
	});
});
