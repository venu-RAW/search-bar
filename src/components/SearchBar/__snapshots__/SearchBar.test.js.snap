// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> component should render the SearchBar 1`] = `
<DocumentFragment>
  <div
    class="searchBar left"
    data-test="searchBar"
    data-testid="searchBar"
  >
    <input
      class="searchInput"
      data-testid="searchBarInput"
      placeholder="Search"
      type="text"
      value=""
    />
    <div
      class="searchBtn"
    >
      <i
        class="fas fa-search"
      />
    </div>
  </div>
</DocumentFragment>
`;
